Let `IO::Memory` not be writable with read-only `Slice`

straight-shoota opened the pull request:
Resolves #10362

bcardiff reviewed (approved):
I nearly forgot about this. Thanks!

straight-shoota commented:
@bcardiff I occasionally look through recently discussed bug issues. This GitHub query helps: [`is:issue is:open sort:updated-desc label:kind:bug -label:community:help-wanted -label:status:someday`](https://github.com/crystal-lang/crystal/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+label%3Akind%3Abug+-label%3Acommunity%3Ahelp-wanted+-label%3Astatus%3Asomeday)

I also have one for non-reviewed PRs: [`is:pr is:open sort:updated-desc -author:@me -reviewed-by:@me -label:status:needs-more-info -label:pr:needs-work  -label:pr:wip `](https://github.com/crystal-lang/crystal/pulls?page=1&q=is%3Apr+is%3Aopen+sort%3Aupdated-desc+-author%3A%40me+-reviewed-by%3A%40me+-label%3Astatus%3Aneeds-more-info+-label%3Apr%3Aneeds-work++-label%3Apr%3Awip).
